cms/core/cache.py

```diff
@@ -36,3 +42,60 @@ def get_default_cache_control_decorator() -> Callable:
     """
     cache_control_kwargs = get_default_cache_control_kwargs()
     return cache_control(**cache_control_kwargs)
+
+
+class ElastiCacheIAMCredentialProvider(redis.CredentialProvider):
+    """A custom redis credential provider to use IAM for authentication.
+
+    https://redis.readthedocs.io/en/stable/examples/connection_examples.html#Connecting-to-a-redis-instance-with-ElastiCache-IAM-credential-provider.
+    """
+
+    # Authentication tokens are only valid for a maximum of 15 minutes.
+    TOKEN_TTL = 900
+
+    def __init__(self, user: str, cluster_name: str, region: str):
+        self.user = user
+        self.cluster_name = cluster_name
+        self.region = region
+
+        session = botocore.session.get_session()
+        self.request_signer = RequestSigner(
+            ServiceId("elasticache"),
+            self.region,
+            "elasticache",
+            "v4",
+            session.get_credentials(),
+            session.get_component("event_emitter"),
+        )
+
+        self.cache_key = f"elasticache_{user}_{cluster_name}_{region}"
+
+        self.connection_url = urlunsplit(
+            SplitResult(
+                scheme="https",
+                netloc=self.cluster_name,
+                path="/",
+                query=urlencode({"Action": "connect", "User": self.user}),
+                fragment="",
+            )
+        )
+
+    def get_credentials(self) -> tuple[str, str]:
+        """Get credentials from IAM."""
+        if (signed_url := caches["memory"].get(self.cache_key)) is None:
```

Comment:
We will have various other uses for in-memory cache. What is your preference? Historically, we have used TTL cache from cachetools; given we have a local memory cache, it seems reasonable to use it. It would be nice to have a decorator to handle the getting/setting to cache using our mem cache.

Reply:
In some cases, `functools.lru_cache` is generally better. We could add a decorator around this, but in most cases we'd probably want the cache centralised through `redis`, rather than in memory.

Reply:
I mean specially for cases where memory is sufficient as we don't want a network call and are happy with cache per process like here.

Reply:
Yeah `lru_cache` doesn't take a TTL :( 

I've added https://github.com/peterbe/django-cache-memoize, and used that for the memory cache.